implementations/rust/message/src/lib.rs

```diff
@@ -0,0 +1,540 @@
+          v.extend_from_slice( ip6.octets().as_ref());
+        },
+      }
+    }
+    fn decode(u: &[u8]) -> Result<(IpAddr, &[u8]), String> {
+      match (HostAddressType::try_from(u[0])?, &u[1..]) {
+        (HostAddressType::Ipv4, addr) => {
+          let ip4 = Ipv4Addr::new(addr[0], addr[1], addr[2], addr[3]);
+          Ok((IpAddr::V4(ip4), &u[5..]))
+        },
+        _ => {
+          Err("".to_string())
+        }
+      }
+    }
+  }
+
+  impl Codec<LocalAddress> for LocalAddress {
+    fn encode(la: &mut LocalAddress, u: &mut Vec<u8>) {
+      u.push(la.length);
+      u.append(&mut la.address);
+    }
+    fn decode(u: &[u8]) -> Result<(LocalAddress, &[u8]), String> {
+      let length =  u[0] as usize;
+      let address =  u[1..length+1].to_vec();
+      Ok((LocalAddress { length: u[0], address }, &u[length+1..]))
+    }
+  }
+
+  /* Routes */
+  #[derive(PartialEq)]
+  #[derive(Debug)]
+  pub struct Route {
+    pub addresses: Vec<Address>,
+  }
+
+  impl Codec<Route> for Route {
+    fn encode(route: &mut Route, u: &mut Vec<u8>) {
+      if route.addresses.len() == 0 {
+        u.push(0 as u8)
+      } else {
+        u.push(route.addresses.len() as u8);
+        for i in (0..route.addresses.len()) {
+          Address::encode(&mut route.addresses[i], u);
+        }
+      }
+    }
+    fn decode(u: &[u8]) -> Result<(Route, &[u8]), String> {
+      let mut route = Route{ addresses: vec![]};
+      let mut w = &u[1..];
+      if 0 == u[0] { return Ok((route, &u[1..])); }
+      for i in 0..u[0] as usize {
+        match Address::decode(w)  {
+          Ok((a, x))  => {
+            route.addresses.push(a);
+            w = x;
+          }
+          Err(s) => {
+            ()
+          }
+        }
+      }
+      Ok((route, w))
+    }
+  }
+
+  #[derive(Debug)]
+  pub enum MessageBody {
+    Ping = 0,
+    Pong = 1,
+    Payload = 2,
+  }
+
+  impl Default for MessageBody {
+    fn default() -> MessageBody {
+      MessageBody::Payload
+    }
+  }
+
+  impl Codec<MessageBody> for MessageBody {
+    fn encode(msg_body: &mut MessageBody, u: &mut Vec<u8>) {
+      match msg_body {
+        MessageBody::Ping => { u.push(MessageBody::Ping as u8); },
+        MessageBody::Pong => { u.push(MessageBody::Pong as u8); },
+        MessageBody::Payload => {}
+      }
+    }
+    fn decode(u: &[u8]) -> Result<(MessageBody, &[u8]), String> {
+      match MessageBody::try_from(u[0])? {
+        MessageBody::Ping => { Ok((MessageBody::Ping, &u[1..])) },
+        MessageBody::Pong => { Ok((MessageBody::Pong, &u[1..])) },
+        _ => Err("Not implemented".to_string())
+      }
+    }
+  }
+
+  impl TryFrom<u8> for MessageBody {
+    type Error = String;
+    fn try_from(data: u8) -> Result<Self, Self::Error> {
+      match data {
+        0 => Ok(MessageBody::Ping),
+        1 => Ok(MessageBody::Pong),
+        _ => Err("Not Implemented".to_string())
+      }
+    }
+  }
+
+  #[allow(arithmetic_overflow)]
+  impl Codec<u16> for u16 {
```

Comment:
usually not a good idea to allow arithmetic_overflow. I wonder which part is causing this error because it's usually `+, -, *, /, %` and I don't see that anywhere except line 309.

Reply:
don't need it after all!